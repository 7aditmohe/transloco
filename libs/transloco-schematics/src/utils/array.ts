export function stringifyList<T>(list: T[], separator = ', '): string {
  return list.map(item => `'${item}'`).join(separator);
}
