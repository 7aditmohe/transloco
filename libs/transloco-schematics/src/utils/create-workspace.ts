import {
  UnitTestTree,
  SchematicTestRunner,
} from '@angular-devkit/schematics/testing';
import { switchMap } from 'rxjs/operators';

export const defaultWorkspaceOptions = {
  name: 'workspace',
  newProjectRoot: 'projects',
  version: '12.1.3',
  minimal: true,
};

export const defaultAppOptions = {
  name: 'bar',
  inlineStyle: false,
  inlineTemplate: false,
  viewEncapsulation: 'Emulated',
  routing: false,
  style: 'css',
  skipTests: false,
};

const defaultLibOptions = {
  name: 'baz',
};

export async function createWorkspace(
  schematicRunner: SchematicTestRunner,
  appTree: UnitTestTree,
  workspaceOptions = defaultWorkspaceOptions,
  appOptions = defaultAppOptions,
  libOptions = defaultLibOptions
) {
  const angularSchematic = '@schematics/angular';
  return schematicRunner
    .runExternalSchematicAsync(angularSchematic, 'workspace', workspaceOptions)
    .pipe(
      switchMap((tree) =>
        schematicRunner.runExternalSchematicAsync(
          angularSchematic,
          'application',
          appOptions,
          tree
        )
      ),
      switchMap((tree) =>
        schematicRunner.runExternalSchematicAsync(
          angularSchematic,
          'library',
          libOptions,
          tree
        )
      )
    )
    .toPromise();
}
