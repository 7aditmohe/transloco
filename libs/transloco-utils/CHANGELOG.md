# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

## [3.0.1](https://github.com/ngneat/transloco/compare/transloco-utils-3.0.0...transloco-utils-3.0.1) (2021-10-28)

### Bug Fixes

- **utils:** 🐛 wrong config search ([29afbf9](https://github.com/ngneat/transloco/commit/29afbf9164607c7e941451031a8f6bf426ff3a6d))
