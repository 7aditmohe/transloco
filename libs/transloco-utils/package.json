{
  "name": "@ngneat/transloco-utils",
  "version": "1.1.0",
  "description": "Transloco utils library",
  "keywords": [
    "angular",
    "angular 2",
    "i18n",
    "transloco",
    "translate",
    "utils",
    "config"
  ],
  "publishConfig": {
    "access": "public"
  },
  "license": "MIT",
  "dependencies": {
    "cosmiconfig": "6.0.0"
  }
}
