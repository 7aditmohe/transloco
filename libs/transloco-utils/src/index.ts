export * from './lib/transloco-utils';
