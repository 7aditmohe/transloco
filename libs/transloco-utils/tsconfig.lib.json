{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../dist/out-tsc",
    "sourceMap": false,
    "declaration": false,
    "module": "commonjs"
  },
  "include": ["**/*.ts"],
  "exclude": ["**/*.spec.ts"]
}
