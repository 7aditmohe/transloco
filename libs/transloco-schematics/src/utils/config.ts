import {getGlobalConfig, TranslocoGlobalConfig} from "@ngneat/transloco-utils";

let config: TranslocoGlobalConfig;

export function getConfig(): TranslocoGlobalConfig {
  if (config) return config;
  config = getGlobalConfig();

  return config;
}
