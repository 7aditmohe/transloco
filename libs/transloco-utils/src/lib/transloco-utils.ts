import { cosmiconfigSync } from 'cosmiconfig';
import * as path from 'path';
import { TranslocoGlobalConfig } from './transloco-utils.types';

export function getGlobalConfig(
  searchPath: string = ''
): TranslocoGlobalConfig {
  const explorer = cosmiconfigSync('transloco');
  const resolvedPath = path.resolve(process.cwd(), searchPath);
  const configSearch = explorer.search(resolvedPath);

  return configSearch ? configSearch.config : {};
}
