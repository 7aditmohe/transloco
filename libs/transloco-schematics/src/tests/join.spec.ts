import { SchematicTestRunner, UnitTestTree } from '@angular-devkit/schematics/testing';
jest.mock('@ngneat/transloco-utils');
import {getGlobalConfig, TranslocoGlobalConfig} from '@ngneat/transloco-utils';
import * as path from 'path';
import { createWorkspace } from '../utils/create-workspace';
import en from './mocks/en';
import es from './mocks/es';
import scopeEn from './mocks/scope-en';
import scopeEs from './mocks/scope-es';

const collectionPath = path.join(__dirname, '../collection.json');

describe('Join', () => {
  const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

  let appTree: UnitTestTree;
  const defaultLang = 'en'
  const options = {
    translationPath: './src/assets/i18n',
    outDir: 'dist-i18n'
  };

  beforeEach(async () => {
    appTree = await createWorkspace(schematicRunner, appTree);
    appTree.create('src/assets/i18n/es.json', JSON.stringify(es));
    appTree.create('src/assets/i18n/en.json', JSON.stringify(en));
    const config: Partial<TranslocoGlobalConfig> = { defaultLang };
    (getGlobalConfig as jest.Mock).mockReturnValue(config);
  });

  describe('default strategy', () => {
    beforeEach(() => {
      appTree.create('src/assets/i18n/scope/en.json', JSON.stringify(scopeEn));
      appTree.create('src/assets/i18n/scope/es.json', JSON.stringify(scopeEs));
    });

    it('should merge translation files that are not the default language to dist', async () => {
      const tree = await schematicRunner.runSchematicAsync('join', options, appTree).toPromise();
      expect(tree.files).toEqual(['/dist-i18n/es.json']);
      expect(tree.files).not.toEqual(['/dist-i18n/en.json']);
    });

    it('should merge translation files including the default language to dist', async () => {
      const tree = await schematicRunner.runSchematicAsync('join', { ...options, includeDefaultLang: true }, appTree).toPromise();
      expect(tree.files).toEqual(['/dist-i18n/es.json', '/dist-i18n/en.json']);
    });

    it('should merge scopes correctly', async () => {
      const tree = await schematicRunner.runSchematicAsync('join', options, appTree).toPromise();

      expect(tree.readContent('/dist-i18n/es.json')).toMatchSnapshot();
    });

    it('should delete output file and pass on rerun', async () => {
      // first run.
      await schematicRunner.runSchematicAsync('join', options, appTree).toPromise();
      // second run.
      const tree = await schematicRunner.runSchematicAsync('join', options, appTree).toPromise();
      expect(tree.files).toEqual(['/dist-i18n/es.json']);
    });

    it(`should take default project's path`, async () => {
      appTree.create('projects/bar/src/assets/i18n/en.json', JSON.stringify(scopeEn));
      appTree.create('projects/bar/src/assets/i18n/es.json', JSON.stringify(scopeEs));

      const tree = await schematicRunner.runSchematicAsync('join', {}, appTree).toPromise();
      expect(tree.files).toEqual(['/dist-i18n/es.json']);
    });

    it('should take specific project path', async () => {
      appTree.create('projects/baz/src/assets/i18n/en.json', JSON.stringify(scopeEn));
      appTree.create('projects/baz/src/assets/i18n/es.json', JSON.stringify(scopeEs));

      const tree = await schematicRunner.runSchematicAsync('join', { project: 'baz' }, appTree).toPromise();
      expect(tree.files).toEqual(['/dist-i18n/es.json']);
    });
  });

  describe.only('scope map strategy', () => {
    function setup(scopePathMap: any = { scope: 'src/app/assets/i18n' }) {
      Object.values(scopePathMap).forEach(path => {
        appTree.create(`${path}/en.json`, JSON.stringify(scopeEn));
        appTree.create(`${path}/es.json`, JSON.stringify(scopeEs));
      });

      (getGlobalConfig as jest.Mock).mockImplementationOnce(() => {
        return { scopePathMap }
      });
    }


    it('should use scope map strategy', async () => {
      setup();
      const tree = await schematicRunner.runSchematicAsync('join', options, appTree).toPromise();
      expect(tree.readContent('/dist-i18n/es.json')).toMatchSnapshot();
    });

    it('should use scope map strategy multi scopes', async () => {
      const scopePathMap = {
        scopeA: 'src/app/assets/i18n/scope1',
        scopeB: 'src/app/assets/i18n/scope2'
      };
      setup(scopePathMap);
      const tree = await schematicRunner.runSchematicAsync('join', options, appTree).toPromise();

      expect(tree.readContent('/dist-i18n/es.json')).toMatchSnapshot();
    });

    it.skip('should use multi projects scopes', async () => {
      const scopePathMap = {
        libA: 'projects/bar/src/assets/i18n',
        libB: 'projects/baz/src/assets/i18n'
      };
      setup(scopePathMap);
      const tree = await schematicRunner.runSchematicAsync('join', options, appTree).toPromise();

      expect(tree.readContent('/dist-i18n/es.json')).toMatchSnapshot();
    });
  });
});
