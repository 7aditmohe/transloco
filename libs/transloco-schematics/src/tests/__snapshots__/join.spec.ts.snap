// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Join scope map strategy should use multi projects scopes 1`] = `
"{
  \\"hello\\": \\"hola\\"
}"
`;

exports[`Join scope map strategy should use multi projects scopes 2`] = `""`;

exports[`Join scope map strategy should use scope map strategy 1`] = `
"{
  \\"hello\\": \\"hola\\"
}"
`;

exports[`Join scope map strategy should use scope map strategy 2`] = `""`;

exports[`Join scope map strategy should use scope map strategy multi scopes 1`] = `
"{
  \\"hello\\": \\"hola\\"
}"
`;

exports[`Join scope map strategy should use scope map strategy multi scopes 2`] = `""`;
